Improvements to inflate chunkset code

nmoinvaz opened the pull request:
This code makes some changes to the chunkset code and also fixes #979.

nmoinvaz committed:
Only need to add rem if it is greater than zero in chunkmemset.

nmoinvaz committed:
Calculate from and out buffer advance only once in chunkcopy.

nmoinvaz committed:
Split out partial chunk copy into its only function.

mtl1979 commented:
I wonder why this is failing several CI tests...

nmoinvaz committed:
Static inline internal chunk memory functions.

nmoinvaz committed:
Store value of out-dist in from variable in chunkmemset_safe.

nmoinvaz commented:
Has something to do with using `chunkcopy_partial` in `chunkmemset`. Memory copy only working when it is one byte at a time.

mtl1979 commented:
@nmoinvaz Instead of reading `len` bytes, it should be reading `dist` bytes and repeatedly writing until it reaches `len` bytes.

If it reads more bytes than `dist`, it will read undefined bytes because those bytes have not been written yet.

nmoinvaz commented:
I ended up re-creating the logic in `inflate_fast`! 

![image](https://user-images.githubusercontent.com/1364432/120874091-ca327e00-c559-11eb-87ce-08259760f216.png)

I suppose that portion should be left as is.

nmoinvaz commented:
Or I can leave `chunkcopy_partial` being called inside `chunkmemset_safe` and take it out of inflate_fast. At least other instances of calling `chunkmemset_safe` will utilize it..

mtl1979 commented:
I think it would be faster to copy `MIN(dist, len)` bytes and then advance `out` by that amount and subtract `len` by that amount.

nmoinvaz commented:
So something like this?
```
       int min_len = MIN(len, dist);
       out = CSUFFIX(chunkcopy_partial(out, from, min_len);
       len -= min_len;
        do {
            *out++ = *from++;
            --len;
        } while (len != 0);
```

mtl1979 commented:
I meant inside the do...while loop... Copying single byte in a loop is very slow...

mtl1979 commented:
```
do {
    int min_len = MIN(len, dist);
    out = CSUFFIX(chunkcopy_partial)(out, from, min_len);
    len -= min_len;
} while (len != 0);
```

It simplifies the code a lot IMHO...

nmoinvaz commented:
I wonder if it would really make it faster though..
If `dist = 1` && `len = 10` it would enter the loop 10 times due to `MIN(len, dist)`.

nmoinvaz commented:
Maybe something like this?
```
int max_len = dist;
do {
    max_len = MIN(max_len, len);
    out = CSUFFIX(chunkcopy_partial)(out, from, max_len);
    len -= max_len;
    max_len += max_len;
} while (len != 0);
```

mtl1979 commented:
At start of loop, it unnecessarily discards the previous value... It should use `MIN` before the loop and on last line of loop...

So...

```
int max_len = MIN(dist, len);
do {
    out = CSUFFIX(chunkcopy_partial)(out, from, max_len);
    len -= max_len;
    max_len = MIN(max_len + max_len, len);
} while (len != 0);
```

nmoinvaz commented:
I ran some benchmarks and it seems the method I have in my commit is faster:
https://www.quick-bench.com/q/DS-sqTR1E8pqgiNyCr3nm5Mw0GQ
See `chunkcopy_a` versus `chunkcopy_b`.

mtl1979 commented:
Still failing a lot of CI tests.

mtl1979 commented on inffast.c:
```diff
@@ -265,10 +265,7 @@ void Z_INTERNAL zng_inflate_fast(PREFIX3(stream) *strm, unsigned long start) {
-                        out = functable.chunkcopy(out, out - dist, len);
-                    else
-                        out = functable.chunkmemset(out, dist, len);
+                    out = functable.chunkmemset_safe(out, dist, len, safe - out);
```
Semantics for checking safe length are not identical in `chunkmemset_safe()` and `chunkcopy_safe()` so they are not interchangeable yet.

mtl1979 commented on chunkset_tpl.h:
```diff
@@ -152,21 +170,25 @@ Z_INTERNAL uint8_t* CHUNKMEMSET(uint8_t *out, unsigned dist, unsigned len) {
 
-Z_INTERNAL uint8_t* CHUNKMEMSET_SAFE(uint8_t *out, unsigned dist, unsigned len, unsigned left) {
+Z_INTERNAL uint8_t* CSUFFIX(chunkmemset_safe)(uint8_t *out, unsigned dist, unsigned len, unsigned left) {
     if (left < (unsigned)(3 * sizeof(chunk_t))) {
```
After this line value of `left` is ignored and as such code assumes `len` is never more than `left`. This is different from `chunkcopy_safe`, which always make sure there is enough room for `len` bytes.

nmoinvaz replied:
Not sure if I need to do anything here.

mtl1979 commented on chunkset_tpl.h:
```diff
@@ -99,6 +114,9 @@ Z_INTERNAL uint8_t* CHUNKMEMSET(uint8_t *out, unsigned dist, unsigned len) {
     if (len < sz) {
+        if (dist >= len || dist >= sizeof(chunk_t)) 
+            return CSUFFIX(chunkcopy_partial)(out, from, len);
+
```
Like mentioned in the comments, if we use minimum of `dist` and `len` for last parameter to `chunkcopy_partial` these two cases can be combined. In "worst" case the loop is gone through only once, in best case the loop is gone through less iterations than when copying just 1 byte in each iteration.

nmoinvaz replied:
Did you see in my comment where I had run the benchmarks?
https://www.quick-bench.com/q/DS-sqTR1E8pqgiNyCr3nm5Mw0GQ
https://www.quick-bench.com/q/q3KyeybzZCj-2JBP5Vezjfu1vPs

`chunkcopy_a`:
```
if (dist >= len || dist >= 32) 
    return chunkcopy_partial(out, from, len);
do {
    *out++ = *from++;
    --len;
} while (len != 0);
return out;
```
`chunkcopy_b`:
```
int max_len = MIN(dist, len);
do {
    out = chunkcopy_partial(out, from, max_len);
    len -= max_len;
    max_len = MIN(max_len + max_len, len);
} while (len != 0);
return out;
```
`chunkcopy_c`:
```
do {
    *out++ = *from++;
    --len;
} while (len != 0);
return out;
```

![image](https://user-images.githubusercontent.com/1364432/120900023-84bc9200-c5e7-11eb-8d2f-17090d637ddc.png)

In this example the `dist=1` and `len=9`. `chunkcopy_b` goes through the loop 4 times instead of 9 with `chunkcopy_c`. However, because of all the extra branches it is actually slower.

mtl1979 replied:
https://www.quick-bench.com/q/8BJmmpeNsflPpY2iJOE9NyLuqT0
https://www.quick-bench.com/q/Xm84euCcPZpYrpBrxtmT4MHQh3U

mtl1979 replied:
https://www.quick-bench.com/q/njzZhYjh-PlG0o3P4S8MvajbYhM
https://www.quick-bench.com/q/AVzYmKDsz2RqzLq99UKkwYs4LJs

nmoinvaz replied:
I didn't see this comment earlier. Good optimization!

mtl1979 replied:
It took me a few attempts to find the sweet spot where it improves performance the most... For short lengths, it's really fast as processor basically runs just one "slow" instruction (repeat move byte). For longer lengths obviously running multiple faster instructions gives better performance.

mtl1979 commented on chunkset_tpl.h:
```diff
@@ -34,37 +63,18 @@ Z_INTERNAL uint8_t* CHUNKCOPY(uint8_t *out, uint8_t const *from, unsigned len) {
-    }
-    return CHUNKCOPY(out, from, len);
+static inline uint8_t* CSUFFIX(chunkcopy_safe_static)(uint8_t *out, uint8_t const *from, unsigned len, uint8_t *safe) {
+    if ((safe - out) < (ptrdiff_t)sizeof(chunk_t))
```
I still think this line is bad. Taking distance of two pointers is never going to give correct length as the range is supposed to be inclusive on both start and end, so `safe` points to last valid memory location, not first out-of-bound location.

nmoinvaz replied:
Not sure if I need to do anything here.

nmoinvaz replied:
I have incorporated some of your changes.

mtl1979 reviewed (changes requested):
Needs addressing comments.

mtl1979 committed:
Fixed off by one error when calculating number of bytes between two pointers.

nmoinvaz commented:
@mtl1979 it appeared the issue was in the `chunkunroll`, it needed to load `from` multiple times, because each time it might have changed when `dist` is small. Also I have incorporated your changes from #982. I was going to originally leave those for your PR, but since you mentioned it I added them here and credited you.

mtl1979 commented:
> I ran some benchmarks and it seems the method I have in my commit is faster:
> https://www.quick-bench.com/q/DS-sqTR1E8pqgiNyCr3nm5Mw0GQ
> See `chunkcopy_a` versus `chunkcopy_b`.

I was actually thinking using memcpy directly in this case... It might be justified for repeated short length copies to reduce number of branches as it might help compiler to optimize complex loop better. When using the hand-written code, it can't inline easily. In best case it would be able to do binary compare with conditional move, but that would require that it knows the variable contents at end of the function are predictable.

nmoinvaz commented on chunkset_tpl.h:
```diff
@@ -2,69 +2,64 @@
-   reliable. */
-Z_INTERNAL uint8_t* CHUNKCOPY(uint8_t *out, uint8_t const *from, unsigned len) {
+/* Copy memory in chunks when possible and avoid writing beyond legal output */
+static inline uint8_t* CSUFFIX(chunkcopy_static)(uint8_t *out, uint8_t const *from, unsigned len, uint8_t *safe) {
```
I have thought about changing the last argument from `safe` to `unsigned left`. Any preference @mtl1979?

mtl1979 replied:
I personally think `left` would simplify the code and reduce risk of off-by-one errors.

nmoinvaz commented:
Changing `inflate_fast` to use all safe functions slows the inflate down by about 16%. https://github.com/zlib-ng/zlib-ng/pull/987/commits/dc80f2a639731e26cf000a509172e23eeb898c64

mtl1979 commented:
32-bit ARM seems to be failing now...

nmoinvaz committed:
Limit number of copy branches based on chunk size in chunkcopy_partial.

nmoinvaz committed:
Must use safe chunk copies due to inflateBack using the same allocation for output and   window. In this instance if too many bytes are written it will not correctly write matches with distances close to the window size.

nmoinvaz committed:
Additional asserts in chunk copy code to prevent invalid parameter values being passed.

nmoinvaz committed:
Use number of bytes left instead of pointer for determine limit of chunk copy. Helps prevent off by one errors.

nmoinvaz committed:
Remove division operation in chunkcopy after changing limit parameter to number of bytes.

nmoinvaz committed:
Use fixed width integer types in chunk functions.

nmoinvaz commented:
## ZLIB-NG (HEAD)

```
 Tool: minigzip-ng.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     48.014%      1.130/1.141/1.149/0.005        0.744/0.757/0.766/0.006      111,512,026
 2     35.391%      1.844/1.866/1.880/0.010        0.762/0.774/0.782/0.005       82,195,416
 3     34.039%      2.261/2.284/2.297/0.009        0.744/0.756/0.764/0.006       79,055,001
 4     32.751%      2.706/2.725/2.739/0.009        0.719/0.735/0.743/0.006       76,065,070
 5     32.484%      2.916/2.937/2.949/0.009        0.714/0.729/0.737/0.006       75,443,521
 6     32.318%      3.520/3.546/3.559/0.010        0.714/0.730/0.738/0.006       75,057,363
 7     32.061%      4.738/4.764/4.778/0.011        0.724/0.733/0.738/0.004       74,461,300
 8     31.978%      7.246/7.285/7.304/0.015        0.718/0.732/0.741/0.006       74,269,032
 9     31.961%      9.823/9.861/9.888/0.016        0.720/0.735/0.743/0.006       74,230,342

 avg1  34.555%                        4.045                          0.742
 tot                               1092.259                        200.452      722,289,071
```

## ZLIB-NG (PR) https://github.com/zlib-ng/zlib-ng/pull/987/commits/5ced1ed7b94083caa9106f3a5a108f2a227448a2

```
 Tool: minigzip-ng-987.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     48.014%      1.122/1.137/1.143/0.005        0.755/0.770/0.779/0.007      111,512,026
 2     35.391%      1.835/1.850/1.860/0.007        0.803/0.815/0.822/0.005       82,195,416
 3     34.039%      2.239/2.271/2.283/0.011        0.791/0.806/0.814/0.006       79,055,001
 4     32.751%      2.697/2.711/2.721/0.006        0.757/0.768/0.773/0.004       76,065,070
 5     32.484%      2.900/2.921/2.930/0.008        0.759/0.772/0.778/0.006       75,443,521
 6     32.318%      3.506/3.530/3.543/0.010        0.754/0.772/0.780/0.007       75,057,363
 7     32.061%      4.754/4.778/4.790/0.009        0.760/0.775/0.782/0.005       74,461,300
 8     31.978%      7.273/7.295/7.308/0.012        0.765/0.777/0.783/0.005       74,269,032
 9     31.961%      9.825/9.868/9.882/0.013        0.765/0.777/0.786/0.006       74,230,342

 avg1  34.555%                        4.040                          0.781
 tot                               1090.842                        210.954      722,289,071
```

Commit https://github.com/zlib-ng/zlib-ng/pull/987/commits/f5325ab3b4ba03da2d713140e18e25f8e147c214 checks to see if the same buffer is being used. Looks like it passes all the unit tests too.

nmoinvaz commented:
## ZLIB-NG HEAD
```
 Tool: minigzip-ng.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     48.014%      1.095/1.107/1.114/0.005        0.741/0.757/0.764/0.006      111,512,026
 2     35.391%      1.824/1.841/1.854/0.008        0.754/0.773/0.781/0.006       82,195,416
 3     34.039%      2.232/2.259/2.275/0.010        0.743/0.752/0.759/0.004       79,055,001
 4     32.751%      2.668/2.692/2.708/0.012        0.715/0.729/0.736/0.005       76,065,070
 5     32.484%      2.882/2.904/2.917/0.009        0.713/0.726/0.733/0.005       75,443,521
 6     32.318%      3.477/3.514/3.530/0.014        0.715/0.727/0.735/0.005       75,057,363
 7     32.061%      4.693/4.733/4.749/0.014        0.718/0.728/0.734/0.004       74,461,300
 8     31.978%      7.171/7.236/7.256/0.020        0.720/0.728/0.735/0.004       74,269,032
 9     31.961%      9.743/9.804/9.828/0.022        0.716/0.725/0.730/0.004       74,230,342

 avg1  34.555%                        4.010                          0.738
 tot                               1082.713                        199.333      722,289,071
```
## ZLIB-NG PR https://github.com/zlib-ng/zlib-ng/pull/987/commits/a0f07eacf3614fde97d58ee1a74b53dea2c70377
```
 Tool: minigzip-ng-987.exe Levels: 1-9
 Runs: 70         Trim worst: 40

 Level   Comp   Comptime min/avg/max/stddev  Decomptime min/avg/max/stddev  Compressed size
 1     48.014%      1.124/1.135/1.142/0.005        0.755/0.763/0.771/0.005      111,512,026
 2     35.391%      1.803/1.840/1.850/0.009        0.792/0.807/0.815/0.006       82,195,416
 3     34.039%      2.248/2.261/2.268/0.005        0.789/0.797/0.804/0.005       79,055,001
 4     32.751%      2.679/2.695/2.710/0.010        0.754/0.765/0.771/0.005       76,065,070
 5     32.484%      2.876/2.903/2.913/0.008        0.751/0.765/0.770/0.004       75,443,521
 6     32.318%      3.482/3.511/3.524/0.012        0.762/0.768/0.774/0.004       75,057,363
 7     32.061%      4.710/4.749/4.764/0.015        0.762/0.769/0.774/0.004       74,461,300
 8     31.978%      7.219/7.254/7.275/0.015        0.761/0.769/0.775/0.004       74,269,032
 9     31.961%      9.759/9.823/9.840/0.019        0.762/0.769/0.774/0.004       74,230,342

 avg1  34.555%                        4.019                          0.775
 tot                               1085.139                        209.189      722,289,071
```
